walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,146 @@
 
 package com.google.sps;
 
+import java.util.Collections;
 import java.util.Collection;
+import java.util.List;
+import java.util.ArrayList;
+import java.util.Comparator;
 
 public final class FindMeetingQuery {
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    
+    // This variable will hold all of the invalid meeting times.
+    // This list is sorted by start times, and has no overlapping times.
+    List<TimeRange> invalidTimeRanges = new ArrayList<>();
+    
+    // This is the list of the attendees which attend the meeting.
+    Collection<String> attendees = request.getAttendees();
+
+    // The duration of the meeting request.
+    int meetingDuration = (int) request.getDuration();
+
+    // Populate invalidTimeRanges variable, and sort by increasing start time.
+    populateInvalidTimeRanges(invalidTimeRanges, events, attendees);
+    Collections.sort(invalidTimeRanges, TimeRange.ORDER_BY_START);
+
+    // Merge conflicting TimeRanges.
+    mergeConflictingTimeRanges(invalidTimeRanges);
+
+    // Get the inverse / valid TimeRanges of the invalidTimeRanges.
```

Comment:
I'd recommend being more clear with what you mean by "inverse". Inverse can also mean the reversed direction, which is what I thought you were doing. This didn't make any sense to me, and it wasn't until I read your code that I understood that you were taking the missing time sections not filled by the invalid time ranges.

This is an excellent example how something makes sense to you because you wrote the code, but doesn't make sense (immediately) to someone new to the code and just reading the comments. It now makes perfect sense to me, but it didn't for a bit.